bftengine/src/bftengine/ReplicaImp.cpp

```diff
@@ -2308,6 +2317,11 @@ void ReplicaImp::onTransferringCompleteImp(SeqNum newStateCheckpoint) {
   }
 }
 
+void ReplicaImp::onSeqNumIsSuperStable(SeqNum newSuperStableSeqNum) {
+  if (controlStateManager_->getStopCheckpointToStopAt() == newSuperStableSeqNum) {
+    if (userRequestsHandler->getControlHandlers()) userRequestsHandler->getControlHandlers()->upgrade();
```

Comment:
The abstraction doesn't seem quite right here. If there is a checkpoint to stop at, then it means an upgrade is expected. But if there is no control handler, then we have a major bug. This should be detectable at compile time. If it's possible to trigger an upgrade, than there must be an upgrade handler. At a minimum, there should be a log message, and likely an exception thrown/assert in this case.

Maybe we need to go back to the drawing board and figure out how we couple the control state to the actions to take based on that control state so that there are no gaps in the semantics.

Reply:
@yontyon and I discussed this. Since the control handlers are all or nothing and implemented by the application, it's possible some applications (like tests) may choose not to implement them. This is totally fine.

Reply:
I'm confused by the naming. In here we pretty much say "super stable CP == upgrade". This might be true for now and in our use-case but what we actually want is to "just get notified" about the super stable CP and *therafter* invoke upgrade. In the current implementation we tie those two together. What if we want to use a super stable checkpoint for something else? Like a general point for synchronization; for backup, for system health reports, to onboard new clients/replicas ...
My suggestion is that each control handler has to provide an `onSuperStableCP` API or similar which then invokes upgrade ... or whatever else.

Reply:
Agreed.
Also, I believe we will have to also write in the reserved pages the reason for stopping at the next checkpoint.
This way, we will be able to distinguish between the different cases.
(see #774 )